libs/models/datapoints.py

```diff
@@ -0,0 +1,400 @@
+    value: float
+
+    @validator("zoneKey")
+    def _validate_zone_key(cls, v: str):
+        if "->" not in v:
+            raise ValueError(f"Not an exchange key: {v}")
+        zone_keys = v.split("->")
+        if zone_keys != sorted(zone_keys):
+            raise ValueError(f"Exchange key not sorted: {v}")
+        if v not in EXCHANGES_CONFIG:
+            raise ValueError(f"Unknown zone: {v}")
+        return v
+
+    @validator("value")
+    def _validate_value(cls, v: float):
+        if abs(v) > 100000:
+            raise ValueError(f"Exchange is implausibly high, above 100GW: {v}")
+        return v
+
+    @staticmethod
+    def create(
+        logger: Logger,
+        zone_key: ZoneKey,
+        datetime: datetime,
+        source: dict,
+        value: float,
+        forecasted: bool = False,
+    ) -> Optional["Exchange"]:
+        try:
+            return Exchange(
+                zoneKey=zone_key,
+                datetime=datetime,
+                source=source,
+                value=value,
+                forecasted=forecasted,
+            )
+        except ValueError as e:
+            logger.error(f"Error creating exchange datapoint {datetime}: {e}")
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "sortedZoneKeys": self.zoneKey,
+            "netFlow": self.value,
+            "source": self.source,
+        }
+
+
+class Generation(Datapoint):
+    value: float
+
+    @validator("value")
+    def _validate_value(cls, v: float):
+        if v < 0:
+            raise ValueError(f"Generation cannot be negative: {v}")
+        if v > 500000:
+            raise ValueError(f"Generation is implausibly high, above 500GW: {v}")
+        return v
+
+    @staticmethod
+    def create(
+        logger: Logger,
+        zone_key: ZoneKey,
+        datetime: datetime,
+        source: dict,
+        value: float,
+        forecasted: bool = False,
+    ) -> Optional["Generation"]:
+        try:
+            return Generation(
+                zoneKey=zone_key,
+                datetime=datetime,
+                source=source,
+                value=value,
+                forecasted=forecasted,
+            )
+        except ValueError as e:
+            logger.error(f"Error creating generation datapoint {datetime}: {e}")
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "zoneKey": self.zoneKey,
+            "generation": self.value,
+            "source": self.source,
+        }
+
+
+class ProductionBreakdown(Datapoint):
+    production: ProductionMix
+    storage: Optional[StorageMix]
+
+    @staticmethod
+    def create(
+        logger: Logger,
+        zone_key: ZoneKey,
+        datetime: datetime,
+        source: dict,
+        production: ProductionMix,
+        storage: Optional[StorageMix] = None,
+        forecasted: bool = False,
+    ) -> Optional["ProductionBreakdown"]:
```

Comment:
Unless we are planning on changing things this also support returning dynamic capacities.

Reply:
Do you have an example of a parser that does that so I could have a look?

Reply:
The CA_AB.py parser returns capacity values like this:
```python
            "capacity": {
                "gas": generation["GAS"]["MC"],
                "hydro": generation["HYDRO"]["MC"],
                "battery storage": generation["ENERGY STORAGE"]["MC"],
                "solar": generation["SOLAR"]["MC"],
                "wind": generation["WIND"]["MC"],
                "biomass": generation["OTHER"]["MC"],
                "unknown": generation["DUAL FUEL"]["MC"],
                "coal": generation["COAL"]["MC"],
            },
            "datetime": get_csd_report_timestamp(response.text),
            "production": {
                "gas": generation["GAS"]["TNG"],
                "hydro": generation["HYDRO"]["TNG"],
                "solar": generation["SOLAR"]["TNG"],
                "wind": generation["WIND"]["TNG"],
                "biomass": generation["OTHER"]["TNG"],
                "unknown": generation["DUAL FUEL"]["TNG"],
                "coal": generation["COAL"]["TNG"],
            },
            "source": URL.netloc,
            "storage": {
                "battery": generation["ENERGY STORAGE"]["TNG"],
            },
            "zoneKey": zone_key,
        },
        ```
In the future I also hope we can have separate capacity parsers for production and exchange if the data is at another endpoint. Might be worth taking into account as well when shaping this?

Reply:
Thanks for pointing this out, I wasn't aware of this.  I'm trying to figure out if we act on this data in the backend or if the yaml config overrides this.
Anyways, I think we should keep this information in this parser for now and therefore add it to the data point structure. Though in the future it could probably become a data point of its own, with its own fetch function.

Reply:
> Thanks for pointing this out, I wasn't aware of this. I'm trying to figure out if we act on this data in the backend or if the yaml config overrides this. Anyways, I think we should keep this information in this parser for now and therefore add it to the data point structure. Though in the future it could probably become a data point of its own, with its own fetch function.

I think we use it for the capacities in the frontend at least but I'm not 100% sure. I'm not aware of any other use case but they might be used in the estimations as well? If not that is something they could be used for in the future if the capacity is updated fairly regular (or change hourly such as for the exchanges).

Reply:
where does this capacity data go? I'm really confused that we have multiple ways of collecting capacity data

Reply:
The capacity that is used should say something like this somewhere in the code (I forgot where):
```python
capacity = parsed_capacity | yaml_capacity
```
So if there is parsed capacity that is what is used and if there is not it falls back to the capacity defined in the config file.

Reply:
I'm still unsure about this. Because the app always use the data available in the flowtraced data. So far I have not been able to find where in the flowtracing pipeline we would load the dynamic capacity instead of the static yaml. So I'm starting to think this data might not be used in the end.

Reply:
I think that as long as this is not clearly resolved it falls outside the scope of this project. Also I'm not sure if the production capacities change on a hour to hour basis

Reply:
Ok so I got insights on where the dynamic capacities are used. I would suggest that to keep this PR small we ignore this issue for now and address it in a later PR. The parsers that we are going to update are not using dynamic capacities for now.

Reply:
Would it make sense to add a comment here about this, to ensure anyone working on parsers before "v2" is implemented will know about this?